### ISSUE TITLE
oxlint: better vue support

### ISSUE THREAD

**Issue by Boshen** [oxc-project/oxc#15761]
## Current Situation

`eslint-plugin-vue` cannot be made fully compatible because it relies on its own compiler, which produces a modified AST for ESLint to traverse.

In Oxc, due to the lack of JavaScript bindings inside Vue templates, rules such as unused variables (and many others) do not work and cannot be implemented.

## Proposal

* `eslint-plugin-vue` will not be a target for full compatibility at this time.
* Implement a custom Vue parser in Oxc that extracts the JavaScript sections from templates.
* Attach these extracted JavaScript fragments to the AST of the `<script>` tag and pass them through the rest of the linter pipeline.

**Comment by overlookmotel** [comment]
An alternative approach would be to add Vue nodes to our AST.

I've not looked deeply at Vue, but supporting Svelte's template syntax (`{#if ...}` etc) would not require too many additions to AST. I'm guessing situation with Vue is probably similar.

One problem with parsing the JS fragments in isolation is that some template syntaxes introduce scopes and symbols. e.g. this introduces a scope, and defines a symbol `item` within it:

```vue
<li v-for="item in items">
  {{ item.message }}
</li>
```

Linter should:

1. Not flag `item` in `item.message` as undefined.
2. Flag `items` as undefined if no binding `items` appears above it in AST.

To do that, it needs not just the bits of JS inside `{{ ... }}`, but also an AST which represents the template syntax, and a scopes tree that matches it.

---

Adding types to the AST may feel like the "nuclear option", but I suspect that any halfway approach will cause other kinds of mess, bugs, and complexity - and ultimately cost us more time than going the whole hog in the first place.

I imagine (hope) that contributors from Vue community (and ditto Svelte etc) would be willing to put in a lot of the work if we agree to support these frameworks as first class citizens.

**Comment by Boshen** [comment]
> An alternative approach would be to add Vue nodes to our AST.

Adding other AST nodes is out of scope for this project.

I'd like to take in an extra dependency in the linter for the transform logic. We should not maintain logic for these other compilers.

**Comment by Boshen** [comment]
We now have a POC: https://github.com/zhiyuanzmj/vue-oxc-parser

We need to somehow integrated into the linter pipeline 

https://github.com/oxc-project/oxc/blob/15d367a8f2c6dbb48398dfbe8e82f307a4194898/crates/oxc_linter/src/service/runtime.rs#L905

**Comment by zhiyuanzmj** [comment]
For semantic:
```rust
let result = VueOxcParser::new(&allocator, &source_text)
    .parse_for_semantic();
let program = result.program;
let errors = result.errors;
```
<img width="2884" height="1574" alt="Image" src="https://github.com/user-attachments/assets/4eeec220-3813-4406-8025-244ec25359ab" />

Playground: https://repl.zmjs.dev/vuejs/vue-oxc-parser?main=src%2FMain.vue

**Comment by kazupon** [comment]
FYI:
@ubugeeei is making a parser (compiler).
https://github.com/ubugeeei/vue-compiler-rs

**Comment by ubugeeei** [comment]
this is my hobby ;D

**Comment by zhiyuanzmj** [comment]
vue-oxc-parser using a vue-parser from https://github.com/HerringtonDarkholme/vue-compiler to generate JSX AST for Oxc, Or you can also generate JS AST for Oxc.

**Comment by ubugeeei** [comment]
Yep! However, that project has been not been active.

So, I tried to rewrite from srcatch as my hobby project.

Feel free for feature requests ;D

**Comment by ubugeeei** [comment]
Hello. As mentioned above, I’m currently implementing a Vue compiler as a hobby project.

While I was thinking about implementing a few linters as well, I came up with an idea.

Similar to how oxlint absorbed tsgolint, what do you think about providing a standard format for consuming lint results from external CLIs via stdout, and making that part of the plugin system?
I haven’t fully thought through the concrete format yet, but I’d be very interested to hear what the oxlint team thinks about this idea.

**Comment by ubugeeei** [comment]
thread with @auvred (Golar)
https://x.com/ubugeeei/status/2010283293414232138?s=20

**Comment by liangmiQwQ** [comment]
I am not the team but I also agree with your idea 😄

- Integrating Vue support into the current oxlint can be pretty difficult. I have made some attempts in private, there are a lot of way to create or use an symbol in vue.

  <details>

  <summary>⬅️ An example to show the difficulty to do the semantic</summary>

  ```vue
  <template>
    <div class="box"> {{ count }} </div>
    <p>{{ msg }}</p>

    <div :key="item.id" v-for="item in list"> <!-- use `item` before it define, but work -->
      {{ item.name }}
    </div>

    <Input v-model:[index].name="name" />
    <Component #default="{ data }"> {{ data }} </Component> <!-- define data there -->
  </template>

  <script lang="ts">
  export default {
    data() {
      return {
        count: 1 // define count there in option api
      }
    }
  }
  </script>

  <script lang="ts" setup>
  import { ref, computed, defineProps } from 'vue';
  
  defineProps<{ msg: string }>() <!-- actually even if we don't define `msg`, it can also be accessed in template -->
  // defineProps({ msg: String })

  const count = ref(2); // define count in composition api
  const fontSize = computed(() => (count.value * 20) + "px")

  const name = ref('Liang Mi');
  
  const index = 1;
  const list = [{name: "Tim", id}, {name: "Jack"}]
  </script>

  <style>
  .box {
    font-size: v-bind(fontSize); /* use the symbol in style with v-bind */
  }
  </style>
  ```

  </details>

  We can transfer it into classic jsx for semantic, but will lost position locations and also hard

  What's more, if the vue support is packed into `oxc_linter`, we are also supposed to fork the transfer and semantic logic to deal with dependencies problem and keep possible breaking changes sync

- And more importantly, as you said, if oxlint does it, you could use this method to support any type of file in the world, and the specific implementation could be finished in any language. It sounds really attractive

**Comment by ubugeeei** [comment]
FYI
https://github.com/ubugeeei/vize/tree/main/crates/vize_patina/src

**Comment by liangmiQwQ** [comment]
❤️ Amazing work! However, I also thought of a problem

- this method might work very well with source file which are totally not related to javascript, or as an extendision rules pack to js/ts files themselves (like tsgolint)

- But for the files like Vue, which partical includes JS and other parts, the difficulty may grow, because we are still supposed to lint the script part which should have same config, rules and result as oxlint. This means we either still generate semantic and let oxlint to check, or call oxlint in the separated vue linter, which may lead into a complex way...

**Comment by ubugeeei** [comment]
Thanks, Uh… yeah, I’m actually working on it right now…!

**Comment by rozsazoltan** [comment]
I am leaving this here just as a note: Vue components imported later via `import()` are incorrectly reported as false-positive errors. While shims-vue would hide the problem, I do not want to mask it. I am looking forward to a fix that properly resolves Vue files as well. For example, this is needed when using async imports:

```js
defineAsyncComponent(() => import('./Header.vue'))
```

```
typescript(TS2307): Cannot find module './Header.vue' or its corresponding type declarations.
    ╭─[packages/module/tests/dev/useDashboard.ts:55:47]
 54 │   {
 55 │     Header: defineAsyncComponent(() => import('./Header.vue')),
    ·                                               ──────────────
 56 │     Counter: defineAsyncComponent(() => import('./Counter.vue')),
    ╰────
```

**Comment by liangmiQwQ** [comment]
I've done some tries to make some improvements to `vue-oxc-parser` originally created by zhiyuanzmj, it now supports reporting parsing errors from `oxc` and filter some warns from `vue-compiler-core`, also created a similar `panicked` logic like oxc_parser.

Besides, I've also done some bug fixes and it can do better in some edge cases

The fork: https://github.com/liangmiQwQ/vue-oxc-toolkit

The next step I plan is creating a prototype to generate `irregular_whitespaces` and `modules_records` that linter required

**Comment by liangmiQwQ** [comment]
I'm making some updates on my Vue POC fork ([vue-oxc-toolkit](https://github.com/liangmiQwQ/vue-oxc-toolkit)):

Now it can generate `irregular_whitespaces` and `module_record` while parsing as a preparation for `import` plugin and `eslint/no-irregular-whitespace` rule.

However, since [vue-compiler-core](https://github.com/HerringtonDarkholme/vue-compiler) doesn't support report it while lexing, we have to iter it after ast generated. Also, The `export` statement of some Vue components are implicit, so the `span` of it is also need discussed, I'm not sure if my current code is correct.

**Comment by liangmiQwQ** [comment]
There are some updates on [vue-oxc-toolkit](https://github.com/liangmiQwQ/vue-oxc-toolkit):

I overrode `v-for`, `v-slot` transform logic and borrowed official Vue compiler's approach, it now can deal with some cases better like `<Comp :key="a" v-for="[a = 1, b] of items" #="user" />` (`:key` called before `v-for`). By the way, now all transform logics run in one single AST traverse, which means probably performance improvements.

I also modified attribute name generate logic for all Vue directives, now it keeps original attribute names without any escape for better linting analyzing. Also, the struct of generated namespace attribute name is more reasonable, too.

Now it's also able to collect all code comments from js/ts parts, which might be helpful for semantic or comments linting

The next step is a bit more complex, I plan to add `<script>` tag transform, including modify it's location ~~and find a way to generate something for some Vue macros like `defineProps` which create new symbols in template.~~[^1] Also includes better error recovery.

[^1]: The official [eslint-plugin-vue](https://github.com/vuejs/eslint-plugin-vue) actually just regards it as normal function call, so there is no need to do that for now.

**Comment by liangmiQwQ** [comment]
Update: After some adjustments, it seems that all tests passed in #19133 😊

**Comment by TobyPerplex** [comment]
Nice, I'm still using eslint alongside oxlint for some vue template rules (see below). Hoping I can say goodbye to eslint when this gets merged.

```mjs
        rules: {
            'vue/block-order': [
                'error',
                {
                    order: ['script', 'template', 'style'],
                },
            ],
            'vue/max-attributes-per-line': [
                'error',
                {
                    singleline: 20,
                    multiline: {
                        max: 1,
                    },
                },
            ],
            'vue/html-self-closing': [
                'error',
                {
                    html: {
                        void: 'always',
                        normal: 'never',
                        component: 'always',
                    },
                    svg: 'always',
                    math: 'always',
                },
            ],
            'vue/no-multiple-template-root': 'off',
            'vue/no-v-html': 'off',
            'vue/padding-line-between-tags': ['error', [{ blankLine: 'consistent', prev: '*', next: '*' }]],
        },
```

**Comment by BryceBarbara** [comment]
@liangmiQwQ do you know the status of the PR review?

**Comment by liangmiQwQ** [comment]
> [@liangmiQwQ](https://github.com/liangmiQwQ) do you know the status of the PR review?

I haven't got PR review yet as I know.

And it's normal because this is a big feature and my approach is just a poc, also probably has some technical limitations (like Vue's JS Plugin or something about [vize](https://github.com/ubugeeei/vize)) and it's not a definite plan. Personally, I also hope @Boshen can review it soon to know whether I'm heading in the right direction before doing more work.

**Comment by ubugeeei** [comment]
I'm working at https://github.com/ubugeeei/vize/pull/115

I may end up requesting a few changes regarding the Oxlint JS plugin.
Mainly, they would be around custom formatting for reports.

**Comment by escaton** [comment]
@ubugeeei the current plan is to support template-related rules of vue plugin via custom js-plugin, right? I suppose this approach does not expose the template ast for other js-plugins to enable custom rules for vue templates, as the eslint-plugin does. Can you help me to understand how theoretically I can write my own rules for vue templates with oxc?

**Comment by ubugeeei** [comment]
My plans are my own and not those of the Oxc team (I’m not part of the Oxc team). 
At the moment, I haven’t started working on enabling users to implement arbitrary rules themselves.

For now, I plan to at least provide configuration options to toggle the rules I currently maintain on or off, but the priority is to deliver something that works first.

Custom rules based on the SFC and Template parser AST from the Vize side will likely come a bit later.

**Comment by mateusfn98** [comment]
Signed

**Comment by TobyPerplex** [comment]
Sealed

**Comment by messenjer** [comment]
RFC: https://github.com/oxc-project/oxc/discussions/21936

**Comment by camc314** [comment]
To bring this issue to it's closure, we are not going to be taking the approach mentioned above.

We are going to be introducing `language plugins` to oxlint. These language plugins allow linting of arbitary files, they can also be passed through the native lint/typecheck plugin if a `transform` is provided.

As mentioned by @ messenjer please see the RFC: #21936 

Thanks!

**Comment by doraemonxxx** [comment]
https://github.com/vbenjs/vue-vben-admin/pull/7945  🥲

**Comment by daniel100097** [comment]
fyi, I've done a prototype for Vue.js by modifying tsgolint and oxlint, using the JS-based Vue.js language tools as reference for the conversion. (runs it as dependency directly)
My company currently tests it internally, maybe not the best implementation, but our typechecks / ESLint type-aware checks went from 20 min down to 30s.

We’ll release patches in coming days for everyone who can’t wait for official implementation. These will serve as stopgap.

**Comment by Norbiros** [comment]
@daniel100097 could you please explain what is the difference between your approach and https://github.com/oxc-project/oxc/pull/24262?

**Comment by daniel100097** [comment]
@Norbiros I used #24262 as the basis for the language plugin system, but I had to add a lot more to support both oxlint’s Vue.js rules and TypeScript rules. fyi I think it would also work with svelte etc

here some ascii diagram on how it works:
 
```
                         ┌──────────────────────┐
                         │   Physical App.vue   │
                         │  User-owned source   │
                         └──────────┬───────────┘
                                    │
                         Vue language adapter
                                    │
              ┌─────────────────────┼─────────────────────┐
              │                     │                     │
              ▼                     ▼                     ▼
    ┌──────────────────┐  ┌──────────────────┐  ┌──────────────────┐
    │  Native Vue AST  │  │ Virtual TS/JS    │  │ Source mappings  │
    │                  │  │ service script   │  │                  │
    │ Elements         │  │                  │  │ Generated ranges │
    │ Directives       │  │ Generated by     │  │        ↕         │
    │ Expressions      │  │ Vue Language     │  │ Original ranges  │
    │ Tokens/comments  │  │ Tools            │  │                  │
    └────────┬─────────┘  └─────────┬────────┘  └─────────┬────────┘
             │                      │                     │
             ▼              ┌───────┴────────┐            │
    ┌──────────────────┐    │                │            │
    │ Vue-aware rules  │    ▼                ▼            │
    │                  │ ┌────────────┐ ┌────────────────┐│
    │ Work directly on │ │ Oxlint     │ │ TypeScript-Go  ││
    │ authored syntax  │ │ Rust rules │ │                ││
    └────────┬─────────┘ │            │ │ Typechecking   ││
             │           │ Fast native│ │ Type-aware     ││
             │           │ linting    │ │ rules/projects ││
             │           └─────┬──────┘ └───────┬────────┘│
             │                 │                │         │
             └─────────────────┴────────────────┘         │
                               │                          │
                               ▼                          │
                    ┌──────────────────────┐              │
                    │ Generated diagnostics│              │
                    │ and proposed fixes   │              │
                    └──────────┬───────────┘              │
                               │                          │
                               └──────────┬───────────────┘
                                          │
                              Map and validate results
                                          │
                         ┌────────────────┴────────────────┐
                         │                                 │
                         ▼                                 ▼
              ┌────────────────────┐            ┌────────────────────┐
              │ Keep               │            │ Suppress           │
              │                    │            │                    │
              │ Safely mapped      │            │ Generated-only     │
              │ diagnostics/fixes  │            │ Vue helper details │
              └─────────┬──────────┘            └────────────────────┘
                        │
                        ▼
              ┌────────────────────┐
              │ User-facing output │
              │                    │
              │ App.vue            │
              │ Original locations │
              │ Safe fixes only    │
              └────────────────────┘
```

**Comment by Togetic** [comment]
### JS-plugin rules can already read Vue templates today — they just can't point at them

A data point from a production Nuxt monorepo (1527 SFCs) that runs oxlint 1.77.0 with JS plugins and still keeps ESLint around **solely** for `eslint-plugin-vue` template rules.

We found that linting Vue templates under oxlint is already possible without native support: a JS-plugin rule can `fs.readFileSync(context.filename)` and parse the full SFC with `@vue/compiler-sfc`, then walk the real template AST. That part works well and we ship rules built this way.

What does not work is **reporting**. oxlint gives JS-plugin rules a script-only view of an SFC, and every way of expressing a template position fails:

| how the rule reports | result on oxlint 1.77.0 |
| --- | --- |
| `loc` with the real template line/column | `RangeError: Line/column pair translates to an out of range offset` |
| `node.range` as absolute **file** offsets | accepted, but the diagnostic renders with **no `line:column` at all** |
| `node.range` as script-relative offsets (negative, since template precedes script) | ``TypeError: `node.range[0]` and `node.range[1]` must be non-negative integers`` |
| `node.range` clamped to `[0, 1]` | renders — on the **script block's first line**, i.e. the wrong place |

Ranges behave as script-relative (oxlint adds the script block's start offset), so the only addressable positions lie inside `<script>`. A template position is therefore representable only when `<script>` precedes `<template>`.

**That ordering is the catch.** In our codebase: **1514 template-first SFCs, 0 script-first.** Template-first is also what `create-vue` scaffolds and what the SFC docs and most style guides show, so the script-first escape hatch is unavailable to a large share of real projects.

Our shipped workaround is to clamp the range to `[0, 1]` and prepend the true location to the message text:

```
Comp.vue:8:1  error  [template 3:15] Raw template text "…" should be wrapped in $t()
```

The squiggle lands on the script block, so editor problem lists and jump-to-error go to the wrong line.

<details>
<summary><b>Minimal repro</b> (oxlint 1.77.0, Node 24.18.0)</summary>

`Comp.vue`
```vue
<template>
    <div>
        <span>raw text</span>
    </div>
</template>

<script setup lang="ts">
const x = 1;
console.log(x);
</script>
```

`plugin.mjs`
```js
import fs from "node:fs";

const probe = (attempt) => ({
    meta: { type: "problem", docs: { description: "", recommended: false }, schema: [], messages: {} },
    create: (context) => ({
        Program() {
            const text = fs.readFileSync(context.filename, "utf8");
            const start = text.indexOf("raw text");                       // in <template>, line 3
            const scriptStart = text.indexOf(">", text.indexOf("<script")) + 1;
            try {
                attempt(context, { start, end: start + 8, scriptStart });
            } catch (e) {
                console.error(`  -> THREW ${e.name}: ${e.message}`);
            }
        },
    }),
});

export default {
    meta: { name: "probe" },
    rules: {
        loc: probe((c) => c.report({
            loc: { start: { line: 3, column: 14 }, end: { line: 3, column: 22 } },
            message: "loc at real template position 3:15",
        })),
        abs: probe((c, i) => c.report({
            node: { type: "X", range: [i.start, i.end] }, message: "absolute file offset",
        })),
        neg: probe((c, i) => c.report({
            node: { type: "X", range: [i.start - i.scriptStart, i.end - i.scriptStart] },
            message: "script-relative offset",
        })),
        zero: probe((c) => c.report({
            node: { type: "X", range: [0, 1] }, message: "[template 3:15] clamped fallback",
        })),
    },
};
```

`.oxlintrc.json`
```json
{
    "jsPlugins": ["./plugin.mjs"],
    "rules": {
        "probe/loc": "error",
        "probe/abs": "error",
        "probe/neg": "error",
        "probe/zero": "error"
    }
}
```

`oxlint --config .oxlintrc.json Comp.vue`
```
  -> THREW RangeError: Line/column pair translates to an out of range offset
  -> THREW TypeError: `node.range[0]` and `node.range[1]` must be non-negative integers
Comp.vue: error probe(abs): absolute file offset
Comp.vue:8:1: error probe(zero): [template 3:15] clamped fallback
```
</details>

#### Why this may be relevant here

Offered as input to the design rather than as a request:

1. Whichever parser ends up winning, if template diagnostics are surfaced through the JS-plugin report path then this span-mapping constraint sits on the critical path.
2. #24262 is actively designing this area (identity span mapping, "keep safely mapped diagnostics / suppress the rest"). Allowing `loc`/`range` to address the **whole file** rather than only the script block — or accepting a file-absolute range — would make userland template rules usable in the interim, and looks considerably smaller than full template support.

Happy to run any candidate branch (#19133 included) against our corpus and report back if that would be useful.

**Comment by Norbiros** [comment]
> We found that linting Vue templates under oxlint is already possible without native support

Exactly. A similar approach is already used by Vize (an experimental Vue toolchain written in Rust). Its oxlint plugin calls Vize's native linter from an oxlint JS plugin, then reports Vize's diagnostics back through oxlint ([source](https://vizejs.dev/guide/oxlint/index.html)). It runs into essentially the same limitations you described, which are tracked separately in https://github.com/oxc-project/oxc/issues/20465.

If https://github.com/oxc-project/oxc/pull/20501 gets merged, it would open up quite a few possibilities. In particular, the idea of running a separate parser + linting engine (either Vize, your approach with `@vue/compiler-sfc` and custom AST walking, or something else entirely) becomes much more practical.

**Comment by justingrant** [comment]
Do any of the tools proposed in this thread work with Vue 2?  Or are they only Vue 3?

> In our codebase: **1514 template-first SFCs, 0 script-first.** Template-first is also what `create-vue` scaffolds and what the SFC docs and most style guides show, so the script-first escape hatch is unavailable to a large share of real projects.

This is the same for ours. A few hundred Vue 2 templates, zero script-first.

**Comment by aentwist** [comment]
Supporting Vue 2 is not reasonable. It has been EOL for 3 years now, with the Vue 3 press release 6 years old. Furthermore `script setup` was backported a long long time ago.

The ordering is an interesting escape hatch. 'Unavailable' is a little strong. Obviously a proper solution should be robust and not require out-of-spec hacks, but just in case there is always AI or the old fashioned way -

Find `<template>(.*)</template>\n+<script>(.*)</script>`
Replace `<script>$2</script>\n\n<template>$1</template>`

Or something like that. Sounds like a review from hell 🤣
